spring-cloud-function-adapters/spring-cloud-function-adapter-gcp/src/test/java/org/springframework/cloud/function/adapter/gcloud/integration/GcfSpringBootHttpRequestHandler2IntegrationTest.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2012-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.function.adapter.gcloud.integration;
+
+import com.google.gson.Gson;
+import org.junit.Rule;
+import org.junit.Test;
+
+import org.springframework.boot.test.web.client.TestRestTemplate;
+import org.springframework.cloud.function.adapter.gcloud.GcfSpringBootHttpRequestHandler2;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.ResponseEntity;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Integration tests for GCF Http Functions.
+ *
+ * @author Daniel Zou
+ * @author Mike Eltsufin
+ */
+public class GcfSpringBootHttpRequestHandler2IntegrationTest {
+
+	private static final int PORT = 7777;
+
+	private static final Gson gson = new Gson();
+
+	@Rule
```

Comment:
I had to do `@Rule` here rather than `@ClassRule` because the context that tells which function to run is cached in `initialize(targetContext)`. But then I ran into a different problem with the `CloudFunctionServer` not shutting down the port.

Reply:
@dzou

Reply:
Maybe try adding a `Thread.sleep(1000)` at the very end of the `after()` method in the `CloudFunctionServer`, it might need some time to shutdown the server before moving onto the next test. If this is the case, then adding a `Thread.sleep` at the end of the before() and after() would be a good enough solution to reduce the flakiness for now.

Reply:
It didn't nothing for me.
https://github.com/GoogleCloudPlatform/functions-framework-java/issues/18